# The Final Project

## Faculty Statement

This lesson is meant to be an easier one to follow after IS 305. It incorporates the basics of how we learned to read and write out files, specifically csv's for this use case. We use the basics as the example and proceed to explore how to use sqllite3 in python as well. We touched sql a tiny bit in class, and I decided I wanted to explore more, as all I know is that it is a commonly used language when it comes to parsing through databases, which happen to be similar to csv’s. I decided to retrieve my csv’s from a discord bot game called Karuta as it’s a game me and my friends have played over many years, so there would be a lot of data stored, and I can explore many things I want too using sql.

## Introduction to Concept

Python is a big programming language, with a lot of variety and options for users to accomplish many tasks. You can program some games, process data, use it to make files, make a script for a common task to repeat over and over, all in a simpler language compared to Java or other coding languages. We also have a very huge and diverse amount of modules or packages from the library than can be imported and used for many different tasks.

SQL, also known as Structured Query Language is another programming language, however, when compared to other languages like Python, instead of having a large amount of versatility for usage in different scenarios, SQL is made for databases. It allows us to make search queries as well as filter, manage, and modify data being stored in a database. It makes it simple for users to process data and search up anything they want. SQLite 3 is a simpler and friendlier version of SQL. It lets the user utilize SQL on a database they make or import right away, compared to needing a server or some other setup used for normal SQL. It still allows us to use basic query searches and make our own database that can be used for personal projects.

The goal for this lesson is to show how we can use python to process data, then proceed to sqlite 3 to make queries and learn more about our data after the cleaning and processing phase. The first goal would be to utilize pathlib, to allow us to access our files, letting us read in data to process. From their, we can clean and choose the data we want overall and proceed to use pathlib to write a new output file, with just the data we want to put into a database. We can then proceed to import sqlite3 and use it and python to convert a csv into a database and explore how to setup and make any queries we want on our database.


## Core Syntax

For this lesson, we will need to be able to use csv, pathlib, and sqlite3. The imports are as follow:

In [ ]:
import csv
import pathlib
import sqlite3

The first thing we need to know is how to read in the csv. This is pretty simple as all we do is use the following template in order to open or read the contents of the file to save the headers and data in our code. The code to do this is:

In [ ]:
with open("file_name", "r", encoding="encoding type") as infile:
    headers, *data = csv.reader(infile)

The r in the code stands for read mode, which allows us to use the reader function in order to save our headers and data in arrays. We define the encoding type for the file to be able to read it in properly, which is usually utf-8. Later on, when we are done cleaning up and saving all the data we want and have our new headers ready, we can use the template again and modify it a bit to write out our new csv file. The code to do this is:

In [ ]:
with open("file_name.csv", "w", encoding="utf-8", newline="") as outfile:
    csvout = csv.writer(outfile)
    csvout.writerow(new_headers)
    csvout.writerows(new_data)

The w is for write, and we specify the new_headers and new_data with the corresponding data. The next step would be for us to setup sqlite3. The first thing we do is use pathlib to make a variable to store our new database files we will make. The code to do this is:

In [ ]:
db_target = pathlib.Path('database_name.db')

Before we work on making our database file, we should make sure that a file with the same name doesnt already exist. It could exist because we have already tried compiling before and saw that we made it incorrectly, so to make sure it comes out properly, we can use the following code to delete the old file if it exists before making a new one:

In [ ]:
if db_target.exists():
    db_target.unlink()

Since we are using sqlite3 in python, we need to setup a connection and cursor to be able to execute SQL commands. To do this, we use the following code:

In [ ]:
conn = sqlite3.connect(db_target)

c = conn.cursor()

Now that we have a cursor setup, we can use the following methods to execute any SQL code we want. The following code has a basic outline of how to make a table and how we can insert data from our csv of data:

In [ ]:
c.execute("CREATE TABLE table_name (col1 text, col2 integer, col3 date);")

c.executemany("INSERT INTO table_name VALUES (?,?,?);", data)

The question marks in the parenthesis after VALUES automatically for each row. We use the execute method when we want to do a single SQL command, such as creating a new table. We use the executemany method when we want to do an SQL command multiple times, such as when we were inserting data from our csv into our database, letting it fill the table in with multiple rows.

When we are done creating our table, we can commit the changes and close the connection in order to create our database file. The code to do this is:

In [ ]:
conn.commit()

conn.close()

When we run a query on our database and want to see the results, we use one of the fetch methods in order to get the results and save them to a variable to print out. The three different fetch methods are fetchone, fetchall, and fetchmany. Fetchone returns one row, fetchall would return all the rows, and fetchmany requires a value to return a certain amount of rows that we the user want. An example of this is:

In [ ]:
c.execute("SELECT * FROM table;")

result_one = c.fetchone() # returns the first row
result_all = c.fetchall() # returns all the rows
result_many = c.fetchmay(5) # returns the first five rows

# To print out the results

print(result_one)

for row in result_all: # or result_many
    print (row)

## Small Example 1

For this example, lets practice opening up a csv file and selecting only certain headers and data that we want and write out the file.

For starters, lets import csv and use the template to open up one of the provided files to load in the headers and data. For this specific example, I will be using the saber_karuta_collection.csv file. The code for this is:

In [1]:
import csv

with open("saber_karuta_collection.csv", "r", encoding='utf-8') as infile:
    headers, *data = csv.reader(infile)

Now that we have read in our headers and data, we can check that they imported properly or just to see what all the headers are if we have never opened the csv before. The code to check the headers is:

In [2]:
print(headers)

['code', 'number', 'edition', 'character', 'series', 'quality', 'obtainedDate', 'obtainedTimestamp', 'burnValue', 'dye.code', 'dye.name', 'frame', 'morphed', 'trimmed', 'tag', 'alias', 'wishlists', 'fights', 'dropQuality', 'dropper', 'grabber', 'guild', 'worker.effort', 'worker.style', 'worker.purity', 'worker.grabber', 'worker.dropper', 'worker.quickness', 'worker.toughness', 'worker.vanity', 'worker.recoveryDate', 'worker.recoveryTimestamp']


To make it easier to see the index value for each column, we can use the following code:

In [3]:
for index, col_name in enumerate(headers):
    print(col_name, index)

code 0
number 1
edition 2
character 3
series 4
quality 5
obtainedDate 6
obtainedTimestamp 7
burnValue 8
dye.code 9
dye.name 10
frame 11
morphed 12
trimmed 13
tag 14
alias 15
wishlists 16
fights 17
dropQuality 18
dropper 19
grabber 20
guild 21
worker.effort 22
worker.style 23
worker.purity 24
worker.grabber 25
worker.dropper 26
worker.quickness 27
worker.toughness 28
worker.vanity 29
worker.recoveryDate 30
worker.recoveryTimestamp 31


We can also check our actual data. We could print out data, but as you can see, it is just the whole csv of data and really hard to read and know where one row starts and one row ends.

In [4]:
# print(data)

To make it easier to go row by row, we can use this code to go row by row, since data is stored as a multidimensional array. This code prints out the whole entire first row(the index is for the first row is 0).

In [5]:
print(data[0])

['v2szs16', '58467', '1', 'Charlie', 'Meiji Tokyo Renka Movie: Yumihari no Serenade', '2', 'Fri, 03 Nov 2023 20:50:11 GMT', '1.69904E+12', '5', '', '', '', 'No', 'No', 'burn', '', '1', '0', '2', '480776931629858823', '480776931629858823', '768148469667201054', '1', 'F', 'S', 'S', 'S', 'A', 'F', 'D', '', '']


Because the data is saved as a multidimensional array, we can even get a specific column from a specific row. For example, if we want the name of the character in the 5th row, we can use the following code to access that information:

In [6]:
print(data[4][3])

Enmu


Just to double check this is correct...

In [7]:
print(data[4])

['v2dlslx', '14458', '3', 'Enmu', 'Demon Slayer: Kimetsu no Yaiba', '2', 'Thu, 02 Nov 2023 18:31:25 GMT', '1.69895E+12', '84', '', '', '', 'No', 'No', 'burn', '', '114', '0', '2', '480776931629858823', '480776931629858823', '768148469667201054', '44', 'F', 'S', 'S', 'S', 'A', 'F', 'D', '', '']


Now that we know how to access our data, the next step is to figure out what columns of data we want. For this example, lets just grab the card code, the print number of the card, what edition the card is, the name of the character on the card, the series the character on the card is from, the current quality of the card, the wishlist number of the card(also known as how many people want that specific character), and the quality the card originally dropped in.

To do this, we can make an array to store the index values of the specific columns of data that we want in our new csv file.

In [8]:
index_values = [0, 1, 2, 3, 4, 5, 16, 18]

Now, we can make our new header array for the final csv file that we will make at the end. To do this, we can make an empty array to store the new headers and loop through our index_values and append to our new headers array the corresponding index value from the original headers array we read in. The code to do this is:

In [9]:
new_headers = []

for i in index_values:
    new_headers.append(headers[i])
    
print(new_headers)

['code', 'number', 'edition', 'character', 'series', 'quality', 'wishlists', 'dropQuality']


Next, we want to make a new array to store the new data that will contain the data from each column for every row. To do this, we can loop through the data, tracking the index and row of each row of data, make a temporary array to store the new row, and use another for loop inside to get the specific columns of data we want for each row to append into our new data array. The code to do this is:

In [10]:
new_data = []

for index, row in enumerate(data):
    new_row = []
    for i in index_values:
        new_row.append(row[i])
    new_data.append(new_row)

print(new_data[4])

['v2dlslx', '14458', '3', 'Enmu', 'Demon Slayer: Kimetsu no Yaiba', '2', '114', '2']


Now that we have our new headers and data, all we have to do is use the template to create our new csv with the data that we want to use. I will be saving the new csv as example1.csv. The code to do this is:

In [11]:
with open("example1.csv", "w", encoding="utf-8", newline="") as outfile:
    csvout = csv.writer(outfile)
    csvout.writerow(new_headers)
    csvout.writerows(new_data)

Now if we check the directory we are working in, we will see a new csv file that has just the data we wanted. We can use this csv for our next example to practice some SQL.

## Small Example 2

For this example, lets use the csv we made in the previous example to practice using SQLite3 and make a new database with a table and import the data from that csv.

For starters, lets import csv, sqlite3, and pathlib so that we can read in our csv file and set up the new database file we will be working with. Don't forget to include the code shown previously, so that if you mess up and want to recreate the file later, it will delete the old one before making a new one. I will name the database file as example2.

In [12]:
import csv
import sqlite3
import pathlib

with open("example1.csv", "r", encoding='utf-8') as infile:
    headers, *data = csv.reader(infile)

db_example = pathlib.Path('example2.db')

if db_example.exists():
    db_example.unlink()

Next, we need to make sure we set up the connection and cursor so that we will be able to execute our SQL code inside of python. The code for this is:

In [13]:
conn = sqlite3.connect(db_example)

c = conn.cursor()

Now that we have everything set up, we need to create a table in SQL to hold our data. Let's name the table example2 and set it up to have the column names be the same as our headers. We have to make sure to specify the type of data that will be contained in these columns. Another thing to notice is that we should also include the phrase IF NOT EXISTS while creating the table. If we run our code to make the table for the first time, it would work just fine. But if we try to run the code again later, it will throw an error as the table already exists.

To check our headers to make sure they are accurate in our SQL table to import the data

In [14]:
print(headers)

['code', 'number', 'edition', 'character', 'series', 'quality', 'wishlists', 'dropQuality']


In [15]:
c.execute("CREATE TABLE IF NOT EXISTS example2 (code text, number integer, edition integer, character text, series text, quality integer, wishlists integer, dropQuality integer);")

In case you mess up the creation of the table, you can't just change the original code to create the table. You would have to delete the table entirely and then recreate the table. The code to do this is the following, and you could just have it before creating the table so that you are always creating the table fresh every time you run your code.

In [16]:
c.execute("DROP TABLE IF EXISTS example2")

c.execute("CREATE TABLE IF NOT EXISTS example2 (code text, number integer, edition integer, character text, series text, quality integer, wishlists integer, dropQuality integer);")

Once you have the table created properly, the next step is to populate the table with the data from our csv. To do this is actually very simple, as SQLite3 is able to handle the data input very easily. Remember to use executemany since we are inserting multiple rows of data. We put 8 question marks as we have 8 columns of data per row. The code to do so is the following:

In [17]:
c.executemany("INSERT INTO example2 VALUES (?, ?, ?, ?, ?, ?, ?, ?);", data)

Congratulations, you have successfully created your own database with a table filled with data from a csv you imported. Unfortunately, if you try to open up the database file after running all the previous code, you will just see an empty table created with the column names but no data. This is because even though we ran the code to insert the data from the csv into the table, we did not actually commit the changes to save the data inside of our table. The code that we need to do so is the following:

In [18]:
conn.commit()

Don't forget to close the connection since we are done now

In [19]:
conn.close()

## The Final Problem

For this final problem, we will take multiple csv's with similar data from different people that have played the card collecting game Karuta. We will extract the data that we want specifically from each person, and add a column to the final csv so that we know who owns that card. We will then follow up by creating a database file using SQL so that we can run different queries to compare the collections. We will have a master table with everyone's card collection data available, as well as make separate tables for each card collection so that we can run certain queries on all the tables to compare results of overall versus individual player preferences or statistics.

First, let's import all our packages needed and open up the csv files with all our players collections data.

In [20]:
import csv
import sqlite3
import pathlib

In [21]:
with open("saber_karuta_collection.csv", "r", encoding='utf-8') as infile:
    saber_headers, *saber_data = csv.reader(infile)

In [22]:
with open("matt_karuta_collection.csv", "r", encoding='utf-8') as infile:
    matt_headers, *matt_data = csv.reader(infile)

In [23]:
with open("gerald_karuta_collection.csv", "r", encoding='utf-8') as infile:
    gerald_headers, *gerald_data = csv.reader(infile)

In [24]:
with open("forming_karuta_collection.csv", "r", encoding='utf-8') as infile:
    forming_headers, *forming_data = csv.reader(infile)

In [25]:
with open("cher_karuta_collection.csv", "r", encoding='utf-8') as infile:
    cher_headers, *cher_data = csv.reader(infile)

In [26]:
with open("boof_karuta_collection.csv", "r", encoding='utf-8') as infile:
    boof_headers, *boof_data = csv.reader(infile)

In [27]:
with open("black_karuta_collection.csv", "r", encoding='utf-8') as infile:
    black_headers, *black_data = csv.reader(infile)

Before proceeding to choose the data we want for our database, lets add an extra column so that we know which person owns the card in their collection.

In [28]:
saber_headers.append("owner")

for row in saber_data:
    row.append("saber")

In [29]:
matt_headers.append("owner")

for row in matt_data:
    row.append("matt")

In [30]:
gerald_headers.append("owner")

for row in gerald_data:
    row.append("gerald")

In [31]:
forming_headers.append("owner")

for row in forming_data:
    row.append("forming")

In [32]:
cher_headers.append("owner")

for row in cher_data:
    row.append("cher")

In [33]:
boof_headers.append("owner")

for row in boof_data:
    row.append("boof")

In [34]:
black_headers.append("owner")

for row in black_data:
    row.append("black")

Just to check we properly added the new column and proper row, we can quickly print out the headers for each person and check the first row.

In [35]:
print(saber_headers)

['code', 'number', 'edition', 'character', 'series', 'quality', 'obtainedDate', 'obtainedTimestamp', 'burnValue', 'dye.code', 'dye.name', 'frame', 'morphed', 'trimmed', 'tag', 'alias', 'wishlists', 'fights', 'dropQuality', 'dropper', 'grabber', 'guild', 'worker.effort', 'worker.style', 'worker.purity', 'worker.grabber', 'worker.dropper', 'worker.quickness', 'worker.toughness', 'worker.vanity', 'worker.recoveryDate', 'worker.recoveryTimestamp', 'owner']


In [36]:
print(saber_data[0])

['v2szs16', '58467', '1', 'Charlie', 'Meiji Tokyo Renka Movie: Yumihari no Serenade', '2', 'Fri, 03 Nov 2023 20:50:11 GMT', '1.69904E+12', '5', '', '', '', 'No', 'No', 'burn', '', '1', '0', '2', '480776931629858823', '480776931629858823', '768148469667201054', '1', 'F', 'S', 'S', 'S', 'A', 'F', 'D', '', '', 'saber']


In [37]:
print(matt_headers)

['code', 'number', 'edition', 'character', 'series', 'quality', 'obtainedDate', 'obtainedTimestamp', 'burnValue', 'dye.code', 'dye.name', 'frame', 'morphed', 'trimmed', 'tag', 'alias', 'wishlists', 'fights', 'dropQuality', 'dropper', 'grabber', 'guild', 'worker.effort', 'worker.style', 'worker.purity', 'worker.grabber', 'worker.dropper', 'worker.quickness', 'worker.toughness', 'worker.vanity', 'worker.recoveryDate', 'worker.recoveryTimestamp', 'owner']


In [38]:
print(matt_data[0])

['v26qz15', '795', '4', 'Muderev', 'The Reincarnation of the Strongest Exorcist in Another World', '4', 'Mon, 06 Nov 2023 03:11:03 GMT', '1699240263663', '160', '', '', '', 'No', 'No', '', '', '1', '0', '4', '289730004600750081', '289730004600750081', '768148469667201054', '85', 'F', 'S', 'S', 'S', 'A', 'F', 'F', '', '', 'matt']


In [39]:
print(gerald_headers)

['code', 'number', 'edition', 'character', 'series', 'quality', 'obtainedDate', 'obtainedTimestamp', 'burnValue', 'dye.code', 'dye.name', 'frame', 'morphed', 'trimmed', 'tag', 'alias', 'wishlists', 'fights', 'dropQuality', 'dropper', 'grabber', 'guild', 'worker.effort', 'worker.style', 'worker.purity', 'worker.grabber', 'worker.dropper', 'worker.quickness', 'worker.toughness', 'worker.vanity', 'worker.recoveryDate', 'worker.recoveryTimestamp', 'owner']


In [40]:
print(gerald_data[0])

['vw86sxs', '60093', '1', 'Hotaru Rindou', 'Qualidea Code', '0', 'Wed, 13 Dec 2023 22:32:18 GMT', '1702506738294', '1', '', '', '', 'No', 'No', '', '', '1', '0', '0', '632803583586205729', '632803583586205729', '768148469667201054', '0', 'F', 'S', 'S', 'S', 'A', 'F', 'D', '', '', 'gerald']


In [41]:
print(forming_headers)

['code', 'number', 'edition', 'character', 'series', 'quality', 'obtainedDate', 'obtainedTimestamp', 'burnValue', 'dye.code', 'dye.name', 'frame', 'morphed', 'trimmed', 'tag', 'alias', 'wishlists', 'fights', 'dropQuality', 'dropper', 'grabber', 'guild', 'worker.effort', 'worker.style', 'worker.purity', 'worker.grabber', 'worker.dropper', 'worker.quickness', 'worker.toughness', 'worker.vanity', 'worker.recoveryDate', 'worker.recoveryTimestamp', 'owner']


In [42]:
print(forming_data[0])

['v2dqnmq', '13635', '3', 'Megumi Fushiguro', 'Jujutsu Kaisen', '1', 'Wed, 01 Nov 2023 20:06:54 GMT', '1698869214710', '44', '', '', '', 'No', 'No', '', '', '2577', '0', '1', '404127890649907204', '226504477161684993', '730537558911484028', '23', 'F', 'S', 'S', 'F', 'S', 'F', 'D', '', '', 'forming']


In [43]:
print(cher_headers)

['code', 'number', 'edition', 'character', 'series', 'quality', 'obtainedDate', 'obtainedTimestamp', 'burnValue', 'dye.code', 'dye.name', 'frame', 'morphed', 'trimmed', 'tag', 'alias', 'wishlists', 'fights', 'dropQuality', 'dropper', 'grabber', 'guild', 'worker.effort', 'worker.style', 'worker.purity', 'worker.grabber', 'worker.dropper', 'worker.quickness', 'worker.toughness', 'worker.vanity', 'worker.recoveryDate', 'worker.recoveryTimestamp', 'owner']


In [44]:
print(cher_data[0])

['vq8f4tf', '55624', '1', 'Abdul Hakim', 'Cowboy Bebop', '2', 'Sat, 29 Jul 2023 00:10:59 GMT', '1690589459958', '7', '', '', '', 'No', 'No', 'kobe', '', '1', '0', '2', '393131628827770890', '393131628827770890', '730537558911484028', '4', 'F', 'S', 'S', 'S', 'A', 'F', 'D', '', '', 'cher']


In [45]:
print(boof_headers)

['code', 'number', 'edition', 'character', 'series', 'quality', 'obtainedDate', 'obtainedTimestamp', 'burnValue', 'dye.code', 'dye.name', 'frame', 'morphed', 'trimmed', 'tag', 'alias', 'wishlists', 'fights', 'dropQuality', 'dropper', 'grabber', 'guild', 'worker.effort', 'worker.style', 'worker.purity', 'worker.grabber', 'worker.dropper', 'worker.quickness', 'worker.toughness', 'worker.vanity', 'worker.recoveryDate', 'worker.recoveryTimestamp', 'owner']


In [46]:
print(boof_data[0])

['v276hz5', '8438', '4', 'Houshou Marine', 'Hololive: Holo no Graffiti', '4', 'Wed, 01 Nov 2023 16:13:49 GMT', '1.69886E+12', '324', '', '', '', 'No', 'No', '', '', '397', '0', '4', '2.47132E+17', '2.47132E+17', '3.43119E+17', '176', 'F', 'S', 'S', 'S', 'S', 'F', 'D', '', '', 'boof']


In [47]:
print(black_headers)

['code', 'number', 'edition', 'character', 'series', 'quality', 'obtainedDate', 'obtainedTimestamp', 'burnValue', 'dye.code', 'dye.name', 'frame', 'morphed', 'trimmed', 'tag', 'alias', 'wishlists', 'fights', 'dropQuality', 'dropper', 'grabber', 'guild', 'worker.effort', 'worker.style', 'worker.purity', 'worker.grabber', 'worker.dropper', 'worker.quickness', 'worker.toughness', 'worker.vanity', 'worker.recoveryDate', 'worker.recoveryTimestamp', 'owner']


In [48]:
print(black_data[0])

['v2866pv', '3011', '2', 'Bruno Bradley', 'The Iceblade Sorcerer Shall Rule the World', '3', 'Sun, 29 Oct 2023 17:46:25 GMT', '1698601585461', '46', '', '', '', 'No', 'No', '', '', '0', '0', '3', '190147150485651457', '190147150485651457', '745688976014180493', '25', 'F', 'S', 'S', 'S', 'A', 'F', 'D', '', '', 'black']


Now we need to determine what data to collect. Since I want to compare different player preferences and statistics, we can collect data that matters in the game, such as worker quality values, the total wishlist collected (can be used to show as an indicator of how big of a collector a player is), and how much a card has been customized(looking at cosmetics such as if a card has been dyed, framed, morphed, trimmed, or given an alias). Let's save our header's in an array called new_headers.

In [49]:
for index, col_name in enumerate(saber_headers):
    print(col_name, index)

code 0
number 1
edition 2
character 3
series 4
quality 5
obtainedDate 6
obtainedTimestamp 7
burnValue 8
dye.code 9
dye.name 10
frame 11
morphed 12
trimmed 13
tag 14
alias 15
wishlists 16
fights 17
dropQuality 18
dropper 19
grabber 20
guild 21
worker.effort 22
worker.style 23
worker.purity 24
worker.grabber 25
worker.dropper 26
worker.quickness 27
worker.toughness 28
worker.vanity 29
worker.recoveryDate 30
worker.recoveryTimestamp 31
owner 32


In [50]:
index_values = [0, 1, 2, 3, 4, 5, 9, 10, 11, 12, 13, 15, 16, 18, 22, 23, 32]

In [51]:
new_headers = []

for i in index_values:
    new_headers.append(saber_headers[i])
    
print(new_headers)

['code', 'number', 'edition', 'character', 'series', 'quality', 'dye.code', 'dye.name', 'frame', 'morphed', 'trimmed', 'alias', 'wishlists', 'dropQuality', 'worker.effort', 'worker.style', 'owner']


Great! Now that we have our headers selected for our csv file, we need to grab the data from each csv and put it into our final data array as well. Let's call our final array new_data.

In [52]:
new_data = []

In [53]:
for index, row in enumerate(saber_data):
    new_row = []
    for i in index_values:
        new_row.append(row[i])
    new_data.append(new_row)

In [54]:
for index, row in enumerate(matt_data):
    new_row = []
    for i in index_values:
        new_row.append(row[i])
    new_data.append(new_row)

In [55]:
for index, row in enumerate(gerald_data):
    new_row = []
    for i in index_values:
        new_row.append(row[i])
    new_data.append(new_row)

In [56]:
for index, row in enumerate(forming_data):
    new_row = []
    for i in index_values:
        new_row.append(row[i])
    new_data.append(new_row)

In [57]:
for index, row in enumerate(cher_data):
    new_row = []
    for i in index_values:
        new_row.append(row[i])
    new_data.append(new_row)

In [58]:
for index, row in enumerate(boof_data):
    new_row = []
    for i in index_values:
        new_row.append(row[i])
    new_data.append(new_row)

In [59]:
for index, row in enumerate(black_data):
    new_row = []
    for i in index_values:
        new_row.append(row[i])
    new_data.append(new_row)

Now lets check to make sure we have all the data collected.

In [60]:
print(len(new_data))

18950


In [61]:
print(new_data[0])

['v2szs16', '58467', '1', 'Charlie', 'Meiji Tokyo Renka Movie: Yumihari no Serenade', '2', '', '', '', 'No', 'No', '', '1', '2', '1', 'F', 'saber']


In [62]:
print(new_data[5000])

['q8hfmb', '6750', '1', 'Suzuki', 'Mayoi Neko Overrun!', '2', '', '', '', 'No', 'No', '', '0', '2', '1', 'F', 'matt']


In [63]:
print(new_data[6000])

['nvmb3r', '334', '2', 'Sherry Polnareff', "JoJo's Bizarre Adventure: Stardust Crusaders", '2', '', '', '', 'No', 'No', '', '1', '2', '25', 'F', 'gerald']


In [64]:
print(new_data[8000])

['clt8w6', '824', '2', 'Rick Dias', 'Mobile Suit Zeta Gundam', '2', '', '', '', 'No', 'No', '', '3', '2', '15', 'F', 'forming']


In [65]:
print(new_data[11000])

['vm6bm5f', '1914', '3', 'Riho Sasaki', 'The Devil Is a Part-Timer!', '4', '', '', '', 'No', 'No', '', '1', '4', '76', 'F', 'cher']


In [66]:
print(new_data[15000])

['13thcn', '995', '3', 'Haruhime Sanjouno', 'Is It Wrong to Try to Pick Up Girls in a Dungeon? II', '3', '', '', '', 'No', 'No', '', '21', '3', '70', 'F', 'black']


In [67]:
print(new_data[18750])

['vbbkdk', '2355', '1', 'Semiramis', 'Fate/Apocrypha', '2', '', '', '', 'No', 'No', '', '35', '2', '23', 'F', 'black']


As we can see, we have all the data, but some columns have blank or empty data values. Let's replace these empty values as None so that when we import our csv into our table, it will become null in SQLite3.

In [68]:
for row in new_data:
    if row[6] == '':
        row[6] = None
        
    if row[7] == '':
        row[7] = None
        
    if row[8] == '':
        row[8] = None
        
    if row[11] == '':
        row[11] = None

In [69]:
print(new_data[0])

['v2szs16', '58467', '1', 'Charlie', 'Meiji Tokyo Renka Movie: Yumihari no Serenade', '2', None, None, None, 'No', 'No', None, '1', '2', '1', 'F', 'saber']


Great! Now we have all our empty values set to None. The final step for this part is to create our final csv file. Let's call this new file master_karuta_collection.csv.

In [70]:
with open("master_karuta_collection.csv", "w", encoding="utf-8", newline="") as outfile:
    csvout = csv.writer(outfile)
    csvout.writerow(new_headers)
    csvout.writerows(new_data)

Now that we have our master csv file with the data we want, the next step is to start using SQLite3. Lets open up our csv file and create our master database file as master.db. Don't forget to setup the connection and cursor so we can use the SQL commands and save our table and changes when we are done.

In [71]:
with open("master_karuta_collection.csv", "r", encoding='utf-8') as infile:
    headers, *data = csv.reader(infile)

db_master = pathlib.Path('master.db')

if db_master.exists():
    db_master.unlink()
    
conn = sqlite3.connect(db_master)

c = conn.cursor()

Let's check our headers and create our master table in our database. Make sure to use execute for creating the table and executemany when importing the data. I will be renaming number as print_number, character as character_name, dye.code as dye_code, dye.name as dye_name, alias as nickname, wishlists as wishlist_count, worker.effort as workEffort, and worker.style as workerStyle.

In [72]:
print(headers)

['code', 'number', 'edition', 'character', 'series', 'quality', 'dye.code', 'dye.name', 'frame', 'morphed', 'trimmed', 'alias', 'wishlists', 'dropQuality', 'worker.effort', 'worker.style', 'owner']


In [73]:
c.execute("DROP TABLE IF EXISTS master")

c.execute("CREATE TABLE IF NOT EXISTS master (code text, print_number integer, edition integer, character_name text, series text, quality integer, dye_code text, dye_name text, frame text, morphed text, trimmed text, nickname text, wishlist_count integer, dropQuality integer, workEffort integer, workerStyle text, owner text);")

In [74]:
c.executemany("INSERT INTO master VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?);", data)

Make sure to commit so that you can open up the table and check that the table is created properly.

In [75]:
conn.commit()

To actually check our data in python, we can use the following code to retrieve each row of data and print it out in python.

In [76]:
c.execute("SELECT * FROM master")

master_table = c.fetchall()

# print(master_table)

Since we have empty values still that aren't showing up as null, we can use SQL to change them into null with the following code:

In [77]:
c.execute("UPDATE master SET dye_code = NULL WHERE dye_code = '';")

In [78]:
c.execute("UPDATE master SET dye_name = NULL WHERE dye_name = '';")

In [79]:
c.execute("UPDATE master SET frame = NULL WHERE frame = '';")

In [80]:
c.execute("UPDATE master SET nickname = NULL WHERE nickname = '';")

In [81]:
conn.commit()

Now when we print out the table again, we should see the empty values replaced with NULL values in the database, or None for Python's case.

In [82]:
c.execute("SELECT * FROM master;")

master_table = c.fetchall()

# print(master_table)

Since we have our master table set up, we can start making comparisons using different search queries. Before we do that however, I will show how we can split up our master table into smaller tables that contain the info for each owner, basically reversing the process of where we combined multiple csv's into one, this time in SQL.

In [83]:
c.execute("DROP TABLE IF EXISTS saber")

c.execute("CREATE TABLE saber AS SELECT * FROM master WHERE owner = 'saber';")

In [84]:
conn.commit()

We can check if it worked by printing the new table and making sure the only value for owner is saber.

In [85]:
c.execute("SELECT * FROM saber;")

saber_table = c.fetchall()

# print(saber_table)

Since it has, we can finish creating the smaller tables for the other players as well.

In [86]:
c.execute("DROP TABLE IF EXISTS matt")
c.execute("CREATE TABLE matt AS SELECT * FROM master WHERE owner = 'matt';")

c.execute("DROP TABLE IF EXISTS gerald")
c.execute("CREATE TABLE gerald AS SELECT * FROM master WHERE owner = 'gerald';")

c.execute("DROP TABLE IF EXISTS forming")
c.execute("CREATE TABLE forming AS SELECT * FROM master WHERE owner = 'forming';")

c.execute("DROP TABLE IF EXISTS cher")
c.execute("CREATE TABLE cher AS SELECT * FROM master WHERE owner = 'cher';")

c.execute("DROP TABLE IF EXISTS boof")
c.execute("CREATE TABLE boof AS SELECT * FROM master WHERE owner = 'boof';")

c.execute("DROP TABLE IF EXISTS black")
c.execute("CREATE TABLE black AS SELECT * FROM master WHERE owner = 'black';")

In [87]:
conn.commit()

Now that we have our master table and smaller tables, we are done with the data cleaning and processing. We can now have fun and explore any queries we want to compare and contrast each player to each other and to the whole dataset as a whole.

The best part of this problem is that we can use sql to make simple search queries and save the results to print out in python instead of doing complicated loops to try and filter data ourselves inside of python. For starters, let's compare everyone's collection size, to see who the largest hoarder is. We can make an array of player names and array to store our query results so it's easy to loop and print out our results.

In [88]:
players = ['saber', 'matt', 'gerald', 'forming', 'cher', 'boof', 'black']
players_collection_sizes = []

In [89]:
c.execute("SELECT COUNT(*) from saber;")

saber_collection_size = c.fetchone()

players_collection_sizes.append(saber_collection_size[0])

In [90]:
c.execute("SELECT COUNT(*) from matt;")

matt_collection_size = c.fetchone()

players_collection_sizes.append(matt_collection_size[0])

In [91]:
c.execute("SELECT COUNT(*) from gerald;")

gerald_collection_size = c.fetchone()

players_collection_sizes.append(gerald_collection_size[0])

In [92]:
c.execute("SELECT COUNT(*) from forming;")

forming_collection_size = c.fetchone()

players_collection_sizes.append(forming_collection_size[0])

In [93]:
c.execute("SELECT COUNT(*) from cher;")

cher_collection_size = c.fetchone()

players_collection_sizes.append(cher_collection_size[0])

In [94]:
c.execute("SELECT COUNT(*) from boof;")

boof_collection_size = c.fetchone()

players_collection_sizes.append(boof_collection_size[0])

In [95]:
c.execute("SELECT COUNT(*) from black;")

black_collection_size = c.fetchone()

players_collection_sizes.append(black_collection_size[0])

In [96]:
for i in range(len(players)):
    print(players[i], "has a collection size of", players_collection_sizes[i])

saber has a collection size of 3962
matt has a collection size of 1485
gerald has a collection size of 2066
forming has a collection size of 3441
cher has a collection size of 1941
boof has a collection size of 272
black has a collection size of 5783


As we can see, boof has the smallest collection size of 272, whereas black has the largest collection size of 5782. Now lets check out who has the highest valued collection in terms of wishlist total. 

In [97]:
player_wishlists = []

In [98]:
c.execute("SELECT SUM(wishlist_count) from saber;")

saber_wishlist = c.fetchone()

player_wishlists.append(saber_wishlist[0])

In [99]:
c.execute("SELECT SUM(wishlist_count) from matt;")

matt_wishlist = c.fetchone()

player_wishlists.append(matt_wishlist[0])

In [100]:
c.execute("SELECT SUM(wishlist_count) from gerald;")

gerald_wishlist = c.fetchone()

player_wishlists.append(gerald_wishlist[0])

In [101]:
c.execute("SELECT SUM(wishlist_count) from forming;")

forming_wishlist = c.fetchone()

player_wishlists.append(forming_wishlist[0])

In [102]:
c.execute("SELECT SUM(wishlist_count) from cher;")

cher_wishlist = c.fetchone()

player_wishlists.append(cher_wishlist[0])

In [103]:
c.execute("SELECT SUM(wishlist_count) from boof;")

boof_wishlist = c.fetchone()

player_wishlists.append(boof_wishlist[0])

In [104]:
c.execute("SELECT SUM(wishlist_count) from black;")

black_wishlist = c.fetchone()

player_wishlists.append(black_wishlist[0])

In [105]:
for i in range(len(players)):
    print(players[i], "has a collection wishlist total of", player_wishlists[i])

saber has a collection wishlist total of 524644
matt has a collection wishlist total of 45752
gerald has a collection wishlist total of 65388
forming has a collection wishlist total of 410193
cher has a collection wishlist total of 191192
boof has a collection wishlist total of 29209
black has a collection wishlist total of 671151


We can see similar results once more with black having the highest amount of wishlist cards in his collection compared to boof having the least wishlist in their collection. However, something to consider would be the number of cards someone has, to see if they have a higher wl average per card, as that could mean they actually have cards that are worth more per wishlist. Let's find out!

In [106]:
for i in range(len(players)):
    print(players[i], "has an average wishlist value per card of", (players_collection_sizes[i]/player_wishlists[i]))

saber has an average wishlist value per card of 0.007551787497808038
matt has an average wishlist value per card of 0.032457597482077286
gerald has an average wishlist value per card of 0.031596011500581146
forming has an average wishlist value per card of 0.008388734083711814
cher has an average wishlist value per card of 0.010152098414159589
boof has an average wishlist value per card of 0.009312198295046047
black has an average wishlist value per card of 0.008616540838052838


This result shows a completly different result, with Matt actually having the highest wishlist average per card and saber having the least.

Another thing we can check easily using SQL is who has the most customized card. This would mean that a card has a quality of 4(the best quality possible), a dye code, a dye name, a frame, is morphed, is trimmed, and has a nickname.

In [107]:
c.execute("SELECT * from master where quality = 4 AND dye_code is not NULL and dye_name is not NULL and frame is not NULL and morphed = 'Yes' and trimmed = 'Yes' AND nickname is not NULL;")
maxed_cards = c.fetchall()
print(maxed_cards)

[('h6llcd', 998, 2, "Jeanne d'Arc (Alter)", 'Fate/Grand Order', 4, '$jwlhd', 'Mystic Girls Night Out', 'blacktiara', 'Yes', 'Yes', 'The Dragon Witch', 366, 3, 350, 'S', 'saber')]


Surprisingly, there is only 1 card out of the 18950 cards in the data that is fully maxed out in terms of customization. This shows that players definetly don't prefer to max out cards, which could be due to the investment required to fully max out a card. We can check this by checking what most player's consider a decent card, that being having a quality of 4, dye code, dye name, and a frame.

In [108]:
c.execute("SELECT * from master where quality = 4 AND dye_code is not NULL and dye_name is not NULL and frame is not NULL;")
decent_cards = c.fetchall()
# print(decent_cards)

There is definitely a lot more cards customized but not fully. Let's see how many in total and the amount from each player.

In [109]:
c.execute("SELECT COUNT(*) from master where quality = 4 AND dye_code is not NULL and dye_name is not NULL and frame is not NULL;")
total_decent_cards = c.fetchall()
print(total_decent_cards[0][0])

191


In [110]:
c.execute("SELECT owner, COUNT(*) as total from master where quality = 4 AND dye_code is not NULL and dye_name is not NULL and frame is not NULL GROUP by owner order by total DESC;")
player_decent_cards = c.fetchall()
print(player_decent_cards)

[('saber', 79), ('black', 54), ('cher', 29), ('forming', 26), ('boof', 3)]


Here we can find a more detailed break down on how many decently customized cards each player has. The query allowed me to order the results in descending order so we can see who has the most first to the least last. We see that Matt and Gerald are missing from the query results, meaning that they have 0 decent cards at all. This could be viewed as them being purely collector players or casual players, that don't care much for the showing off of cards other than the fact that they have one, even if it is uncustomized

Overall, this final problem has allowed us to explore how to combine multiple csv's into one and load it into SQL. We were able to clean and process the data that we wanted and load it into a table for further investigation. We also learned how to break down a huge table into smaller, more categorized table, in our case being for each player. This could easily have been done since we have the csv files for each individual, but in the case we only had a master csv, this is useful to know to help look at a target person in our case. Finally, we were able to do some sample queries to analyze and see the difference in player collection's and their choice on how they customize their cards, being fully customized, decently customized, or not at all. There are definitely a lot more queries that can be done, and I encourage you all to try and make your own from the SQL you have learned here and more you can learn online.

Oh, and don't forget to commit and close your connection once you are done! :)

In [111]:
conn.commit()
conn.close()